### PR TITLE
Port the script to Python3 and fix some issues

### PR DESCRIPTION
**Pull request by marxin**
and I recommend removal of the generated files, it's generating very many git changes..

### CODE REVIEW

**Review by davidmalcolm — APPROVED** [review]
Gah; the "review changes" option seems to apply to the whole pull request, rather than to the individual commits.

Port texi2rst to Python3: Looks good

**Review by davidmalcolm — COMMENTED** [review]
Use Sphinx theme sphinx_rtd_theme: OK

File: texi2rst.py [hunk 1]
```diff
@@ -1216,7 +1216,8 @@ def __init__(self, table_element, debug):
             for idx in range(self.num_columns):
                 column = []
                 for row in comp.rows:
-                    column.append(row.entries[idx])
+                    if idx < len(row.entries):
+                        column.append(row.entries[idx])
```
Comment (davidmalcolm) [inline texi2rst.py:1230]: Why did you need this?  Can we have a unit test please?

**Review by davidmalcolm — COMMENTED** [review]
"Clean up from generated files": the reason I had these in the tree was to make it easy to compare the effect of a change to the script on the generated .rst files.

Can we have a "make" target to do something similar, perhaps?

**Review by davidmalcolm — COMMENTED** [review]
" Add index.rst": what's the purpose of this?  Is it to ensure the creation of the indices and tables?

**Review by davidmalcolm — COMMENTED** [review]
"Add index.rst" : OK

**Review by davidmalcolm — COMMENTED** [review]
"Fix index link": OK

File: conf.py [hunk 2]
```diff
@@ -41,7 +41,7 @@
 
 # General information about the project.
 project = u'gcc'
-copyright = u'2021, gcc peeps'
+copyright = u'2021, GCC'
```
Comment (davidmalcolm) [inline conf.py:44]: Copyright is the Free Software Foundation

File: conf.py [hunk 3]
```diff
@@ -184,7 +184,7 @@
 # (source start file, target name, title, author, documentclass [howto/manual]).
 latex_documents = [
   ('index', 'gcc.tex', u'gcc Documentation',
-   u'gcc peeps', 'manual'),
```
Comment (davidmalcolm) [inline conf.py:187]: Authors should presumably be something like "GCC Contributors"

**Review by davidmalcolm — COMMENTED** [review]
"Port tests to Python 3": LGTM

File: texi2rst.py [hunk 4]
```diff
@@ -425,12 +425,16 @@ class OptionRefFixer(NoopVisitor):
...
         # Conditionally retain options (or else they will be
         # stripped at output)
-        def previsit_element(self, element):
+        def previsit_element(self, element, parent):
             if element.kind == 'option':
                 firstchild = element.children[0]
                 if isinstance(firstchild, Text):
                     if firstchild.data.startswith('-'):
                         element.rst_kind = InlineMarkup('option')
+                        if firstchild.data.endswith('='):
```
Comment (davidmalcolm) [inline texi2rst.py:435]: Maybe a comment here, giving an example?

File: texi2rst.py [hunk 5]
```diff
@@ -1666,6 +1670,13 @@ def test_valid_option_ref(self):
         out = self.make_rst_string(doc)
         self.assertEqual(u':option:`--some-opt`', out)
 
+    def test_option_ref_with_var(self):
+        xml_src = '<para>This is similar to how <option>-Walloca-larger-than=</option><var>byte-size</var> works.</para>'
+        doc = from_xml_string(xml_src)
+        doc = fixup_option_refs(doc)
+        out = self.make_rst_string(doc)
+        self.assertEqual('This is similar to how :option:`-Walloca-larger-than` = byte-size works.\n\n', out)
```
Comment (davidmalcolm) [inline texi2rst.py:1684]: Should the .rst output do something to the {{{= byte-size}}} rather than having it as plain text?  It's enclosed in a <var> element; seems a shame to just drop it on the floor.

**Review by davidmalcolm — CHANGES_REQUESTED** [review]
Approval was just for first patch; gahh; marking as "request changes".

File: texi2rst.py [hunk 6]
```diff
@@ -1677,6 +1683,13 @@ def test_option_ref_with_var(self):
         out = self.make_rst_string(doc)
         self.assertEqual('This is similar to how :option:`-Walloca-larger-than` = byte-size works.\n\n', out)
 
+    def test_option_ref_with_var2(self):
+        xml_src = '<para>See also <option>-Walloca-larger-than=<var>byte-size</var></option>.</para>'
+        doc = from_xml_string(xml_src)
+        doc = fixup_option_refs(doc)
+        out = self.make_rst_string(doc)
+        self.assertEqual('See also :option:`-Walloca-larger-than` = byte-size.\n\n', out)
```
Comment (davidmalcolm) [inline texi2rst.py:1691]: Similar here; would be better for the `byte-size` to have some .rst markup

**Comment by marxin** [comment]
> Gah; the "review changes" option seems to apply to the whole pull request, rather than to the individual commits.

Yep, that's how GitHub pull requests work. That's my mistake that I collected multiple commits together. I'm going  to address the comments.

File: texi2rst.py (hunk 1 above)
Comment (marxin) [reply texi2rst.py:1220]: I'll do it in a different pull request, I accidentally pushed a work-in-progress version.

File: conf.py (hunk 2 above)
Comment (marxin) [reply conf.py]: Done.

File: conf.py (hunk 3 above)
Comment (marxin) [reply conf.py:187]: Yep, fixed now.

File: texi2rst.py (hunk 5 above)
Comment (marxin) [reply texi2rst.py]: Yep, it's work-in-progress..

File: texi2rst.py (hunk 6 above)
Comment (marxin) [reply texi2rst.py]: Yes, will do a separate pull request once I'm happy with the generated `rst` files.

**Comment by marxin** [comment]
> " Add index.rst": what's the purpose of this? Is it to ensure the creation of the indices and tables?

Yes, the files was created by you: `gcc documentation master file, created by sphinx-quickstart...`

File: texi2rst.py (hunk 1 above)
Comment (davidmalcolm) [reply texi2rst.py:1220]: I think this hunk isn't quite resolved yet

**Review by davidmalcolm — CHANGES_REQUESTED** [review]
Looks good, apart from the  "Fix table parsing in x86 Operand Modifiers." patch

File: texi2rst.py (hunk 1 above)
Comment (marxin) [reply texi2rst.py:1220]: You're right. I've added a test-case for it.

File: texi2rst.py [hunk 7]
```diff
@@ -2599,7 +2600,30 @@ def test_multitable_without_header(self):
...
+</para></entry><entry command="tab" spaces=" "><para><code>rax</code>
+</para></entry></row><row><entry command="item" spaces=" "><para><code>Q</code>
+</para></entry><entry command="tab" spaces=" "><para>print the opcode suffix of q.
+</para></entry><entry command="tab" spaces=" "><para><code>%Q0</code>
+</para></entry><entry command="tab" spaces=" "><para><code>q</code>
+</para></entry><entry command="tab">
+</entry></row></tbody></multitable>'''
+        tree = from_xml_string(xml_src)
+        tree = convert_to_rst(tree, self.ctxt)
+        out = self.make_rst_string(tree)
```
Comment (davidmalcolm) [inline texi2rst.py:2627]: Looks like you're trying to suppress a crash here and verify it doesn't crash, but can this also have a `self.assertEqual(` on the resulting value of `out`?  i.e. what .rst does it actually generate for this case, and is it sane?

**Review by davidmalcolm — CHANGES_REQUESTED** [review]

File: texi2rst.py (hunk 7 above)
Comment (marxin) [reply texi2rst.py:2639]: I've just added a test-case for it. Yes, the output seems sane.

**Review by davidmalcolm — APPROVED** [review]

**Comment by davidmalcolm** [comment]
Thanks!
